x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/ip/IpPrefix.java

```diff
@@ -0,0 +1,224 @@
+        );
+        this.ipField = ipField;
+        this.prefixLengthV4Field = prefixLengthV4Field;
+        this.prefixLengthV6Field = prefixLengthV6Field;
+    }
+
+    public static IpPrefix readFrom(PlanStreamInput in) throws IOException {
+        return new IpPrefix(in.readSource(), in.readExpression(), in.readExpression(), in.readOptionalNamed(Expression.class));
+    }
+
+    public static void writeTo(PlanStreamOutput out, IpPrefix ipPrefix) throws IOException {
+        out.writeSource(ipPrefix.source());
+        List<Expression> fields = ipPrefix.children();
+        assert fields.size() == 2 || fields.size() == 3;
+        out.writeExpression(fields.get(0));
+        out.writeExpression(fields.get(1));
+        out.writeOptionalWriteable(fields.size() == 3 ? o -> out.writeExpression(fields.get(2)) : null);
+    }
+
+    public Expression ipField() {
+        return ipField;
+    }
+
+    public Expression prefixLengthV4Field() {
+        return prefixLengthV4Field;
+    }
+
+    public Expression prefixLengthV6Field() {
+        return prefixLengthV6Field;
+    }
+
+    @Override
+    public boolean foldable() {
+        return Expressions.foldable(children());
+    }
+
+    @Override
+    public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        var ipEvaluatorSupplier = toEvaluator.apply(ipField);
+        var prefixLengthV4EvaluatorSupplier = toEvaluator.apply(prefixLengthV4Field);
+
+        if (prefixLengthV6Field != null) {
+            var prefixLengthV6EvaluatorSupplier = toEvaluator.apply(prefixLengthV6Field);
+
+            return new IpPrefixEvaluator.Factory(
+                source(),
+                ipEvaluatorSupplier,
+                prefixLengthV4EvaluatorSupplier,
+                prefixLengthV6EvaluatorSupplier,
+                context -> new BytesRef(new byte[16])
+            );
+        }
+
+        return new IpPrefixOnlyV4Evaluator.Factory(
+            source(),
+            ipEvaluatorSupplier,
+            prefixLengthV4EvaluatorSupplier,
+            context -> new BytesRef(new byte[16])
+        );
+    }
+
+    @Evaluator(extraName = "OnlyV4")
+    static BytesRef process(BytesRef ip, int prefixLengthV4, @Fixed(includeInToString = false, build = true) BytesRef scratch) {
+        return process(ip, prefixLengthV4, 128, scratch);
+    }
+
+    @Evaluator
+    static BytesRef process(
+        BytesRef ip,
+        int prefixLengthV4,
+        int prefixLengthV6,
+        @Fixed(includeInToString = false, build = true) BytesRef scratch
+    ) {
+        if (ip.length != 16 || prefixLengthV4 < 0 || prefixLengthV4 > 32 || prefixLengthV6 < 0 || prefixLengthV6 > 128) {
+            return null;
+        }
+
+        boolean isIpv4 = Arrays.compareUnsigned(ip.bytes, 0, IPV4_PREFIX.length, IPV4_PREFIX, 0, IPV4_PREFIX.length) == 0;
+
+        if (isIpv4) {
+            if (prefixLengthV4 == 32) {
```

Comment:
I probably wouldn't optimize for either of these cases. It's just not going to be common for folks to mask and IP with "all of it". I don't think it hurts anything though.

Reply:
Was thinking more in the "Masking v4 but not v6". But definitely looks like an even weirder case...
Going to remove them